File: routes/culture.py
# 屈氏文化展示（参考 ly-web 屈氏文化研究会内容）
from flask import Blueprint, render_template, request, jsonify
from models import db, GenealogyMain, GenealogyGeneration, FamilyMember, find_relationship

bp = Blueprint('culture', __name__, url_prefix='/culture')


@bp.route('/')
def index():
    """屈氏文化：起源、郡望堂号、关于我们"""
    return render_template('qu_culture.html')


@bp.route('/zibei')
def zibei():
    """字辈查询：按辈字搜索所属族谱"""
    keyword = request.args.get('q', '').strip()
    results = []
    all_genealogies = []

    genealogies = GenealogyMain.query.order_by(
        GenealogyMain.surname, GenealogyMain.region
    ).all()

    for g in genealogies:
        gens = GenealogyGeneration.query.filter_by(genealogy_id=g.id).order_by(
            GenealogyGeneration.sort_order
        ).all()
        if not gens:
            continue
        chars = ''.join(gen.character for gen in gens)
        entry = {
            'genealogy': g,
            'generations': gens,
            'chars_str': chars,
        }
        all_genealogies.append(entry)
        if keyword and keyword in chars:
            matched_positions = []
            for gen in gens:
                if keyword in gen.character:
                    matched_positions.append(gen.sort_order + 1)
            entry['matched_positions'] = matched_positions
            results.append(entry)

    return render_template(
        'zibei_search.html',
        keyword=keyword,
        results=results,
        all_genealogies=all_genealogies,
    )


@bp.route('/contact')
def contact():
    """联系我们"""
    return render_template('contact.html')


# ---------- 亲缘关系查询 ----------

@bp.route('/relationship', methods=['GET', 'POST'])
def relationship():
    """亲缘关系查询"""
    genealogies = GenealogyMain.query.order_by(
        GenealogyMain.surname, GenealogyMain.genealogy_name
    ).all()

    result = None
    form = {
        'genealogy_id': request.form.get('genealogy_id', type=int) or request.args.get('genealogy_id', type=int),
        'name_a': (request.form.get('name_a') or '').strip(),
        'name_b': (request.form.get('name_b') or '').strip(),
        'id_a': request.form.get('id_a', type=int),
        'id_b': request.form.get('id_b', type=int),
    }

    if request.method == 'POST' and form['genealogy_id'] and form['name_a'] and form['name_b']:
        gid = form['genealogy_id']

        if form['id_a']:
            member_a = FamilyMember.query.filter_by(id=form['id_a'], genealogy_id=gid).first()
        else:
            member_a = FamilyMember.query.filter_by(name=form['name_a'], genealogy_id=gid).first()

        if form['id_b']:
            member_b = FamilyMember.query.filter_by(id=form['id_b'], genealogy_id=gid).first()
        else:
            member_b = FamilyMember.query.filter_by(name=form['name_b'], genealogy_id=gid).first()

        members_a = FamilyMember.query.filter_by(name=form['name_a'], genealogy_id=gid).all()
        members_b = FamilyMember.query.filter_by(name=form['name_b'], genealogy_id=gid).all()

        if not members_a:
            result = {'error': f'未找到成员「{form["name_a"]}」'}
        elif not members_b:
            result = {'error': f'未找到成员「{form["name_b"]}」'}
        elif (len(members_a) > 1 and not form['id_a']) or (len(members_b) > 1 and not form['id_b']):
            result = {
                'disambiguate': True,
                'members_a': members_a if len(members_a) > 1 and not form['id_a'] else None,
                'members_b': members_b if len(members_b) > 1 and not form['id_b'] else None,
                'selected_a': member_a,
                'selected_b': member_b,
            }
        else:
            if not member_a:
                member_a = members_a[0]
            if not member_b:
                member_b = members_b[0]
            rel = find_relationship(member_a, member_b)
            if rel:
                result = {**rel, 'member_a': member_a, 'member_b': member_b}
            else:
                result = {
                    'error': f'在此族谱中未找到「{member_a.name}」与「{member_b.name}」之间的亲缘关系',
                    'member_a': member_a,
                    'member_b': member_b,
                }

    return render_template(
        'relationship.html',
        genealogies=genealogies,
        result=result,
        form=form,
    )


@bp.route('/api/members')
def api_members():
    """成员名称自动补全 API"""
    gid = request.args.get('genealogy_id', type=int)
    q = request.args.get('q', '').strip()
    if not gid or not q:
        return jsonify([])
    members = FamilyMember.query.filter(
        FamilyMember.genealogy_id == gid,
        FamilyMember.name.contains(q),
    ).limit(20).all()
    seen = {}
    for m in members:
        key = m.name
        if key not in seen:
            seen[key] = []
        seen[key].append({
            'id': m.id,
            'name': m.name,
            'generation': m.generation_number,
            'birth_date': m.birth_date or '',
            'father': m.father.name if m.father else '',
        })
    results = []
    for name, items in seen.items():
        if len(items) == 1:
            results.append(items[0])
        else:
            for item in items:
                item['display'] = f"{item['name']}（第{item['generation'] or '?'}世" \
                                  + (f"，父：{item['father']}" if item['father'] else '') + '）'
                results.append(item)
    return jsonify(results)

File: packages/server/src/routes/culture.ts
import { Router, Request, Response } from 'express';
import { getDb } from '../db/connection.js';
import { findRelationship } from '../services/relationship.js';

const router: Router = Router();

router.get('/zibei', (req: Request, res: Response) => {
  try {
    const q = (req.query.q as string)?.trim() ?? '';

    const db = getDb();
    const genealogies = db.prepare(`
      SELECT * FROM genealogy_main
      ORDER BY surname, region
    `).all() as Record<string, unknown>[];

    const results: Record<string, unknown>[] = [];

    for (const g of genealogies) {
      const gens = db.prepare(`
        SELECT * FROM genealogy_generation
        WHERE genealogy_id = ?
        ORDER BY sort_order ASC
      `).all(g.id) as { character: string; sort_order: number }[];

      if (gens.length === 0) continue;

      const charsStr = gens.map((gen) => gen.character).join('');
      const entry: Record<string, unknown> = {
        genealogy: g,
        generations: gens,
        chars_str: charsStr,
      };

      if (q && charsStr.includes(q)) {
        const matchedPositions: number[] = [];
        for (const gen of gens) {
          if (gen.character.includes(q)) {
            matchedPositions.push(gen.sort_order + 1);
          }
        }
        entry.matched_positions = matchedPositions;
        results.push(entry);
      } else if (!q) {
        results.push(entry);
      }
    }

    res.status(200).json(results);
  } catch (err) {
    console.error('Zibei search error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.post('/relationship', (req: Request, res: Response) => {
  try {
    const { genealogy_id, name_a, name_b, id_a, id_b } = req.body ?? {};

    if (!genealogy_id || !name_a || !name_b) {
      res.status(400).json({ error: 'genealogy_id, name_a, and name_b are required' });
      return;
    }

    const gid = parseInt(String(genealogy_id), 10);
    if (isNaN(gid)) {
      res.status(400).json({ error: 'Invalid genealogy_id' });
      return;
    }

    const db = getDb();
    const genealogy = db.prepare('SELECT id FROM genealogy_main WHERE id = ?').get(gid);
    if (!genealogy) {
      res.status(404).json({ error: 'Genealogy not found' });
      return;
    }

    let memberA: Record<string, unknown> | null = null;
    let memberB: Record<string, unknown> | null = null;

    if (id_a) {
      memberA = db.prepare('SELECT * FROM family_member WHERE id = ? AND genealogy_id = ?').get(id_a, gid) as Record<string, unknown> | undefined ?? null;
    }
    if (!memberA) {
      memberA = db.prepare('SELECT * FROM family_member WHERE name = ? AND genealogy_id = ?').get(name_a, gid) as Record<string, unknown> | undefined ?? null;
    }

    if (id_b) {
      memberB = db.prepare('SELECT * FROM family_member WHERE id = ? AND genealogy_id = ?').get(id_b, gid) as Record<string, unknown> | undefined ?? null;
    }
    if (!memberB) {
      memberB = db.prepare('SELECT * FROM family_member WHERE name = ? AND genealogy_id = ?').get(name_b, gid) as Record<string, unknown> | undefined ?? null;
    }

    const membersA = db.prepare('SELECT * FROM family_member WHERE name = ? AND genealogy_id = ?').all(name_a, gid) as Record<string, unknown>[];
    const membersB = db.prepare('SELECT * FROM family_member WHERE name = ? AND genealogy_id = ?').all(name_b, gid) as Record<string, unknown>[];

    if (membersA.length === 0) {
      res.status(200).json({ error: `Member "${name_a}" not found` });
      return;
    }
    if (membersB.length === 0) {
      res.status(200).json({ error: `Member "${name_b}" not found` });
      return;
    }

    const needDisambiguateA = membersA.length > 1 && !id_a;
    const needDisambiguateB = membersB.length > 1 && !id_b;

    if (needDisambiguateA || needDisambiguateB) {
      res.status(200).json({
        disambiguate: true,
        members_a: needDisambiguateA ? membersA : null,
        members_b: needDisambiguateB ? membersB : null,
        selected_a: memberA,
        selected_b: memberB,
      });
      return;
    }

    const ma = memberA ?? membersA[0];
    const mb = memberB ?? membersB[0];
    const idA = ma.id as number;
    const idB = mb.id as number;

    const result = findRelationship(idA, idB);
    if (result) {
      res.status(200).json({
        relation: result.relation,
        lca: result.lca,
        up_a: result.upA,
        up_b: result.upB,
        chain_a: result.chainA,
        chain_b: result.chainB,
        member_a: ma,
        member_b: mb,
      });
    } else {
      res.status(200).json({
        error: `No relationship found between "${ma.name}" and "${mb.name}" in this genealogy`,
        member_a: ma,
        member_b: mb,
      });
    }
  } catch (err) {
    console.error('Relationship error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;
